File: public/python/044-course-schedule-ii.py
# LeetCode Problem #210: Course Schedule II
# Difficulty: Medium
# Link: https://leetcode.com/problems/course-schedule-ii/

from typing import List
from collections import deque


# ─────────────────────────────────────────────
# APPROACH 1: Kahn's Algorithm (BFS Topological Sort) | O(V+E) time | O(V+E) space
# EXPLAIN: Enqueue nodes with in-degree 0; each dequeued node goes into the result; cycle ↔ not all nodes processed.
# WHEN: Iterative and intuitive; most common interview answer for topological ordering.

def findOrder_bfs(numCourses: int, prerequisites: List[List[int]]) -> List[int]:
    adj = [[] for _ in range(numCourses)]
    indegree = [0] * numCourses
    for a, b in prerequisites:
        adj[b].append(a)
        indegree[a] += 1
    q = deque(i for i in range(numCourses) if indegree[i] == 0)
    result = []
    while q:
        u = q.popleft()
        result.append(u)
        for v in adj[u]:
            indegree[v] -= 1
            if indegree[v] == 0:
                q.append(v)
    return result if len(result) == numCourses else []


# ─────────────────────────────────────────────
# APPROACH 2: DFS Topological Sort | O(V+E) time | O(V+E) space
# EXPLAIN: Post-order DFS appends each node after all its successors; reverse gives topological order.
# WHEN: Recursive formulation; natural connection between DFS post-order and topological sort.

def findOrder_dfs(numCourses: int, prerequisites: List[List[int]]) -> List[int]:
    adj = [[] for _ in range(numCourses)]
    for a, b in prerequisites:
        adj[b].append(a)
    state = [0] * numCourses  # 0=unvis, 1=visiting, 2=done
    order = []

    def dfs(u):
        if state[u] == 1: return False
        if state[u] == 2: return True
        state[u] = 1
        for v in adj[u]:
            if not dfs(v): return False
        state[u] = 2
        order.append(u)
        return True

    if not all(state[i] == 2 or dfs(i) for i in range(numCourses)):
        return []
    return order[::-1]


# ─────────────────────────────────────────────
# APPROACH 3: Iterative DFS | O(V+E) time | O(V+E) space
# EXPLAIN: Explicit stack DFS avoids call-stack overflow for large graphs.
# WHEN: Safe for very large course graphs; same output as Approach 2.

def findOrder_iterative(numCourses: int, prerequisites: List[List[int]]) -> List[int]:
    adj = [[] for _ in range(numCourses)]
    for a, b in prerequisites:
        adj[b].append(a)
    state = [0] * numCourses
    order = []
    for start in range(numCourses):
        if state[start] != 0:
            continue
        stack = [(start, 0)]  # (node, neighbor_index)
        while stack:
            u, ni = stack[-1]
            if ni == 0:
                if state[u] == 1: return []
                if state[u] == 2: stack.pop(); continue
                state[u] = 1
            if ni < len(adj[u]):
                stack[-1] = (u, ni + 1)
                v = adj[u][ni]
                if state[v] == 0:
                    stack.append((v, 0))
                elif state[v] == 1:
                    return []
            else:
                state[u] = 2
                order.append(u)
                stack.pop()
    return order[::-1]


# ─────────────────────────────────────────────
# APPROACH 4: Kahn's with Min-Heap | O(V log V + E) time | O(V+E) space
# EXPLAIN: Replace FIFO queue with min-heap to produce lexicographically smallest valid order.
# WHEN: When a deterministic smallest ordering is required among valid topological sorts.

def findOrder_pq(numCourses: int, prerequisites: List[List[int]]) -> List[int]:
    import heapq
    adj = [[] for _ in range(numCourses)]
    indegree = [0] * numCourses
    for a, b in prerequisites:
        adj[b].append(a)
        indegree[a] += 1
    heap = [i for i in range(numCourses) if indegree[i] == 0]
    heapq.heapify(heap)
    result = []
    while heap:
        u = heapq.heappop(heap)
        result.append(u)
        for v in adj[u]:
            indegree[v] -= 1
            if indegree[v] == 0:
                heapq.heappush(heap, v)
    return result if len(result) == numCourses else []


# ─────────────────────────────────────────────
# APPROACH 5: DFS with Finish Time | O(V log V + E) time | O(V+E) space
# EXPLAIN: Record the DFS finish timestamp of each node; sort descending to obtain topological order.
# WHEN: Educational — connects topological sort to Tarjan's DFS finish-time concept.

def findOrder_finish_time(numCourses: int, prerequisites: List[List[int]]) -> List[int]:
    adj = [[] for _ in range(numCourses)]
    for a, b in prerequisites:
        adj[b].append(a)
    state = [0] * numCourses
    finish = [0] * numCourses
    timer = [0]

    def dfs(u):
        if state[u] == 1: return False
        if state[u] == 2: return True
        state[u] = 1
        for v in adj[u]:
            if not dfs(v): return False
        state[u] = 2
        finish[u] = timer[0]
        timer[0] += 1
        return True

    if not all(state[i] == 2 or dfs(i) for i in range(numCourses)):
        return []
    return sorted(range(numCourses), key=lambda x: -finish[x])


# Made with Bob

File: public/java/044-course-schedule-ii.java
/*
 * LeetCode Problem #210: Course Schedule II
 * Difficulty: Medium
 * Link: https://leetcode.com/problems/course-schedule-ii/
 */
import java.util.*;

public class Solution {

    // APPROACH 1: Kahn's Algorithm (BFS Topological Sort) | O(V+E) time | O(V+E) space
    // EXPLAIN: Process nodes with in-degree 0; each processed node goes into the result; if all processed, return result, else empty.
    public int[] findOrder(int n, int[][] prereqs) {
        List<List<Integer>> adj = buildAdj(n, prereqs);
        int[] indegree = new int[n];
        for (int[] p : prereqs) indegree[p[0]]++;
        Queue<Integer> q = new LinkedList<>();
        for (int i = 0; i < n; i++) if (indegree[i] == 0) q.offer(i);
        int[] result = new int[n];
        int idx = 0;
        while (!q.isEmpty()) {
            int u = q.poll();
            result[idx++] = u;
            for (int v : adj.get(u)) if (--indegree[v] == 0) q.offer(v);
        }
        return idx == n ? result : new int[]{};
    }

    // APPROACH 2: DFS Topological Sort | O(V+E) time | O(V+E) space
    // EXPLAIN: Post-order DFS; nodes added after all descendants — reverse to get topological order.
    public int[] findOrderDFS(int n, int[][] prereqs) {
        List<List<Integer>> adj = buildAdj(n, prereqs);
        int[] state = new int[n]; // 0=unvis, 1=visiting, 2=done
        List<Integer> order = new ArrayList<>();
        for (int i = 0; i < n; i++)
            if (state[i] == 0 && !dfs(i, adj, state, order)) return new int[]{};
        Collections.reverse(order);
        return order.stream().mapToInt(Integer::intValue).toArray();
    }
    private boolean dfs(int u, List<List<Integer>> adj, int[] state, List<Integer> order) {
        if (state[u] == 1) return false;
        if (state[u] == 2) return true;
        state[u] = 1;
        for (int v : adj.get(u)) if (!dfs(v, adj, state, order)) return false;
        state[u] = 2;
        order.add(u);
        return true;
    }

    // APPROACH 3: Iterative DFS | O(V+E) time | O(V+E) space
    // EXPLAIN: Explicit stack DFS avoids call-stack overflow for large graphs.
    public int[] findOrderIterative(int n, int[][] prereqs) {
        List<List<Integer>> adj = buildAdj(n, prereqs);
        int[] state = new int[n];
        List<Integer> order = new ArrayList<>();
        for (int start = 0; start < n; start++) {
            if (state[start] != 0) continue;
            Deque<int[]> stack = new ArrayDeque<>();  // [node, neighborIdx]
            stack.push(new int[]{start, 0});
            while (!stack.isEmpty()) {
                int[] top = stack.peek();
                int u = top[0], ni = top[1];
                if (ni == 0) {
                    if (state[u] == 1) return new int[]{};
                    if (state[u] == 2) { stack.pop(); continue; }
                    state[u] = 1;
                }
                if (ni < adj.get(u).size()) {
                    top[1]++;
                    int v = adj.get(u).get(ni);
                    if (state[v] == 0) stack.push(new int[]{v, 0});
                    else if (state[v] == 1) return new int[]{};
                } else {
                    state[u] = 2;
                    order.add(u);
                    stack.pop();
                }
            }
        }
        Collections.reverse(order);
        return order.stream().mapToInt(Integer::intValue).toArray();
    }

    // APPROACH 4: Kahn's with Priority Queue | O(V log V + E) time | O(V+E) space
    // EXPLAIN: Min-heap instead of FIFO queue produces lexicographically smallest valid order.
    public int[] findOrderPQ(int n, int[][] prereqs) {
        List<List<Integer>> adj = buildAdj(n, prereqs);
        int[] indegree = new int[n];
        for (int[] p : prereqs) indegree[p[0]]++;
        PriorityQueue<Integer> pq = new PriorityQueue<>();
        for (int i = 0; i < n; i++) if (indegree[i] == 0) pq.offer(i);
        int[] result = new int[n];
        int idx = 0;
        while (!pq.isEmpty()) {
            int u = pq.poll();
            result[idx++] = u;
            for (int v : adj.get(u)) if (--indegree[v] == 0) pq.offer(v);
        }
        return idx == n ? result : new int[]{};
    }

    // APPROACH 5: DFS with Finish Time | O(V log V + E) time | O(V+E) space
    // EXPLAIN: Assign finish timestamps; sort nodes by decreasing finish time for topological order.
    private int timer;
    public int[] findOrderFinishTime(int n, int[][] prereqs) {
        List<List<Integer>> adj = buildAdj(n, prereqs);
        int[] state = new int[n], finish = new int[n];
        timer = 0;
        for (int i = 0; i < n; i++)
            if (state[i] == 0 && !dfsFinish(i, adj, state, finish)) return new int[]{};
        Integer[] ids = new Integer[n];
        for (int i = 0; i < n; i++) ids[i] = i;
        Arrays.sort(ids, (a, b) -> finish[b] - finish[a]);
        return Arrays.stream(ids).mapToInt(Integer::intValue).toArray();
    }
    private boolean dfsFinish(int u, List<List<Integer>> adj, int[] state, int[] finish) {
        if (state[u] == 1) return false;
        if (state[u] == 2) return true;
        state[u] = 1;
        for (int v : adj.get(u)) if (!dfsFinish(v, adj, state, finish)) return false;
        state[u] = 2; finish[u] = timer++;
        return true;
    }

    private List<List<Integer>> buildAdj(int n, int[][] prereqs) {
        List<List<Integer>> adj = new ArrayList<>();
        for (int i = 0; i < n; i++) adj.add(new ArrayList<>());
        for (int[] p : prereqs) adj.get(p[1]).add(p[0]);
        return adj;
    }
}

// Made with Bob
